client/network/src/protocol/notifications/behaviour.rs

```diff
@@ -2070,3 +2073,1985 @@ impl NetworkBehaviour for Notifications {
+		};
+
+		notif.inject_connection_established(&peer, &conn, &connected, None, 0usize);
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Disabled { .. })));
+
+		// manually add backoff for the entry
+		if let Some(PeerState::Disabled { ref mut backoff_until, .. }) =
+			notif.peers.get_mut(&(peer, set_id))
+		{
+			*backoff_until =
+				Some(Instant::now().checked_add(std::time::Duration::from_secs(5)).unwrap());
+		}
+
+		notif.inject_connection_closed(
+			&peer,
+			&conn,
+			&connected,
+			NotifsHandlerProto::new(vec![]).into_handler(&peer, &connected),
+			0usize,
+		);
+
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Backoff { .. })));
+
+		// attempt to connect the backed-off peer and verify that the request is pending
+		notif.peerset_report_connect(peer, set_id);
+		assert!(std::matches!(
+			notif.peers.get(&(peer, set_id)),
+			Some(&PeerState::PendingRequest { .. })
+		));
+
+		notif.peerset_report_connect(peer, set_id);
+	}
+
+	#[test]
+	#[should_panic]
+	#[cfg(debug_assertions)]
+	fn peerset_report_disconnect_with_incoming_peer() {
+		let (mut notif, _peerset) = development_notifs();
+		let peer = PeerId::random();
+		let set_id = sc_peerset::SetId::from(0);
+		let conn = ConnectionId::new(0usize);
+		let connected = ConnectedPoint::Listener {
+			local_addr: Multiaddr::empty(),
+			send_back_addr: Multiaddr::empty(),
+		};
+
+		notif.inject_connection_established(&peer, &conn, &connected, None, 0usize);
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Disabled { .. })));
+
+		// remote opens a substream, verify that peer state is updated to `Incoming`
+		notif.inject_event(peer, conn, NotifsHandlerOut::OpenDesiredByRemote { protocol_index: 0 });
+
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Incoming { .. })));
+
+		notif.peerset_report_disconnect(peer, set_id);
+	}
+
+	#[test]
+	#[should_panic]
+	#[cfg(debug_assertions)]
+	fn peerset_report_accept_incoming_peer() {
+		let (mut notif, _peerset) = development_notifs();
+		let peer = PeerId::random();
+		let conn = ConnectionId::new(0usize);
+		let set_id = sc_peerset::SetId::from(0);
+		let connected = ConnectedPoint::Listener {
+			local_addr: Multiaddr::empty(),
+			send_back_addr: Multiaddr::empty(),
+		};
+
+		notif.inject_connection_established(&peer, &conn, &connected, None, 0usize);
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Disabled { .. })));
+
+		// remote opens a substream, verify that peer state is updated to `Incoming`
+		notif.inject_event(peer, conn, NotifsHandlerOut::OpenDesiredByRemote { protocol_index: 0 });
+		assert!(std::matches!(notif.peers.get(&(peer, set_id)), Some(&PeerState::Incoming { .. })));
+
+		assert!(std::matches!(
+			notif.incoming[0],
+			IncomingPeer { alive: true, incoming_id: sc_peerset::IncomingIndex(0), .. },
+		));
+
+		notif.peers.remove(&(peer, set_id));
```

Comment:
Why manually remove peer if we are going to accept it (as the test name suggests)?

Reply:
Is it testing panicking in case inconsistent internal state is found?

Reply:
yeah, there was some code path (I can't exactly remember which) that assumes the peer is present and panics otherwise